### PR TITLE
Cache director registry requests

### LINKED ISSUE #878: Getting services from backend takes too long

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?

- optimized calls to registry by using asyncio.gather which increase speed massively (divide by max factor 10)
- added caching mechanism that caches every 15minutes

## Related issue number
fixes #878 
links #885 
links #880 

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] **Runs in the swarm**
- [x] If you design a new module, add your user to .github/CODEOWNERS

### CODE REVIEW

File: services/director/src/simcore_service_director/registry_proxy.py [hunk 1]
```diff
@@ -125,58 +206,39 @@ class ServiceType(enum.Enum):
...
+        prefixes.append(_get_prefix(ServiceType.DYNAMIC))
+    if service_type in [ServiceType.COMPUTATIONAL, ServiceType.ALL]:
+        prefixes.append(_get_prefix(ServiceType.COMPUTATIONAL))
+    repos = [x for x in repos if str(x).startswith(tuple(prefixes))]
+    _logger.debug("retrieved list of repos : %s", repos)
+    
     # only list as service if it actually contains the necessary labels
+    tasks = [get_repo_details(repo) for repo in repos]
+    results = await asyncio.gather(*tasks, return_exceptions=True)
     services = []
```
Comment (mguidon) [inline services/director/src/simcore_service_director/registry_proxy.py]: That is neat!

File: services/director/src/simcore_service_director/registry_proxy.py [hunk 2]
```diff
@@ -1,25 +1,70 @@
...
 
 DEPENDENCIES_LABEL_KEY = 'simcore.service.dependencies'
 
-NUMBER_OF_RETRIEVED_REPOS = 5
-NUMBER_OF_RETRIEVED_TAGS = 5
+NUMBER_OF_RETRIEVED_REPOS = 50
+NUMBER_OF_RETRIEVED_TAGS = 50
 
 _logger = logging.getLogger(__name__)
+_registry_requests_cache = {}
```
Comment (pcrespov) [inline services/director/src/simcore_service_director/registry_proxy.py]: Avoid singleton for registry. Add it instead in the app config SEE https://aiohttp.readthedocs.io/en/stable/web_advanced.html#data-sharing-aka-no-singletons-please

File: services/director/src/simcore_service_director/registry_proxy.py [hunk 3]
```diff
@@ -1,25 +1,70 @@
...
     DYNAMIC = "dynamic"
 
+class TaskState(enum.IntEnum):
+    STARTING = 0
+    RUNNING = 1
+    FAILED = 2
+    STOPPED = 3
+
+async def registry_caching_task(app: web.Application):
+    _logger.info("starting registry caching task")
```
Comment (pcrespov) [inline services/director/src/simcore_service_director/registry_proxy.py]: what you call a "caching task"  is in reality a cache cleaner ... this task does not cache anything. i would rename it for clarity

File: services/director/src/simcore_service_director/registry_proxy.py [hunk 4]
```diff
@@ -65,17 +116,22 @@ class ServiceType(enum.Enum):
     raise exceptions.RegistryConnectionError("Unknown registry authentification type: {}".format(url))
 
 async def _registry_request(path: URL, method: str ="GET") -> Tuple[Dict, Dict]:
+    cache_key = "{}_{}".format(path, method)
```
Comment (pcrespov) [inline services/director/src/simcore_service_director/registry_proxy.py]: I wonder why you did not create this with a decorator
Why you only use it here and not in ``_auth_registry_request`` as well?

File: services/director/src/simcore_service_director/registry_proxy.py [hunk 5]
```diff
@@ -102,13 +163,35 @@ class ServiceType(enum.Enum):
...
     return repos_list
 
-async def _get_image_details(image_key: str, image_tag: str) -> Dict:
+async def list_image_tags(image_key: str) -> List[Dict]:
+    _logger.debug("listing image tags in %s", image_key)
+    image_tags = []
+    # get list of image tags
+    loop = True
+    request = URL("v2/{}/tags/list?n={}".format(image_key, NUMBER_OF_RETRIEVED_TAGS))
+    while loop:
```
Comment (pcrespov) [inline services/director/src/simcore_service_director/registry_proxy.py]: IMHO this reads better that using the variable ``loop`` which in this context typically refers to the event loop
```python
header = ""
while "Link" in header:
  header = ...
```

File: services/director/tests/conftest.py [hunk 6]
```diff
@@ -70,3 +71,18 @@ def user_id():
 @pytest.fixture
 def project_id():
     yield "some_project_id"
+
+def pytest_addoption(parser):
```
Comment (pcrespov) [inline services/director/tests/conftest.py]: Nice

File: services/director/src/simcore_service_director/registry_proxy.py [hunk 7]
```diff
@@ -125,58 +208,39 @@ class ServiceType(enum.Enum):
...
+    if service_type in [ServiceType.DYNAMIC, ServiceType.ALL]:
+        prefixes.append(_get_prefix(ServiceType.DYNAMIC))
+    if service_type in [ServiceType.COMPUTATIONAL, ServiceType.ALL]:
+        prefixes.append(_get_prefix(ServiceType.COMPUTATIONAL))
+    repos = [x for x in repos if str(x).startswith(tuple(prefixes))]
+    _logger.debug("retrieved list of repos : %s", repos)
 
     # only list as service if it actually contains the necessary labels
+    tasks = [get_repo_details(repo) for repo in repos]
+    results = await asyncio.gather(*tasks, return_exceptions=True)
```
Comment (pcrespov) [inline services/director/src/simcore_service_director/registry_proxy.py]: very nice!

File: services/director/src/simcore_service_director/registry_proxy.py (hunk 3 above)
Comment (sanderegg) [reply services/director/src/simcore_service_director/registry_proxy.py]: that is not true. the call to list_services repopulates the cache.

File: services/director/src/simcore_service_director/registry_proxy.py (hunk 4 above)
Comment (sanderegg) [reply services/director/src/simcore_service_director/registry_proxy.py]: never did a decorator before, that is probably why.
_auth_registry_request is a subfunction that gets called by registry_request when authentication is necessary, therefore no need to cache so deep.

File: services/director/src/simcore_service_director/registry_proxy.py (hunk 5 above)
Comment (sanderegg) [reply services/director/src/simcore_service_director/registry_proxy.py]: goog catch! I actually need a do...while type of loop. so changed it to while True, if condition false break instead.

File: services/director/src/simcore_service_director/registry_proxy.py (hunk 3 above)
Comment (pcrespov) [reply services/director/src/simcore_service_director/registry_proxy.py]: aha! my bad! :-)

File: services/director/src/simcore_service_director/registry_proxy.py (hunk 2 above)
Comment (sanderegg) [reply services/director/src/simcore_service_director/registry_proxy.py]: this was painful... but I think I got it right __and__ with a decorator...
all tests are modified accordingly...

**Pull request merged**